### DISCUSSION TITLE
ActionText

### DISCUSSION THREAD

**Discussion by chrpai** [wixtoolset/issues#7320] [Questions]
This is an MSI bug impacts WiX that I first noticed about 10 years. ago  According to the MSI SDK:

https://learn.microsoft.com/en-us/windows/win32/msi/actiontext-table

> If the ActionText table is not populated, the installer loads localized strings for the language specified by the [ProductLanguage](https://learn.microsoft.com/en-us/windows/win32/msi/productlanguage) property.

Well, this used to be true... about 10+ years ago.  But at some point, win7 or 8 I don't recall,  a break was introduced into MSI that caused an ugly UI experience like this:

![image](https://user-images.githubusercontent.com/43178814/228354203-29a3956f-6b42-49aa-8896-e85b3b6f8f20.png)

So when I build WiX with WiXUI I get no ActionText table and ProductLanguage set to 1033.  I see the above.

If I add EnsureTable ActionText I get some rows for Wix4 custom actions but none of the standard MSI ones because well you aren't supposed to need them.  

If I build an InstallShield MSI they ship the full ActionText table cause that's the kind of bloat they like to do.  Except in this case it works.  If I export their ActionText table to IDT format and import it into my MSI I suddenly get:

![image](https://user-images.githubusercontent.com/43178814/228355438-e219ca18-83e8-4199-9e7f-378b1f01597e.png)

Has anyone else ever seen this?  I wanted to discuss before filing an issue.  It would be nice if WiX UI populated the table based on the language.  Even nicer would be if it only populated the needed rows for the actions included in the MSI.

**Comment by chrpai** [comment]
I did some digging and realized that I must have solved this years ago.  One of my installers has UIRef WixUI_ErrorProgressText in it.

Now I'm wondering why referencing standard UI such as WixUI FeatureTree doesn't pull this in automatically.

**Reply by robmen** [reply]
Pulling it in bloats an MSI since the Windows Installer is supposed to show messages correctly. So, we don't do so automatically but provide the mechanism if you want to manually. If we automatically pulled it in, you would not be able to not do so. :smile:

Fundamentally, Windows should stop breaking the UI here.

**Comment by chrpai** [comment]
The bloat, for english after building, is 28KB. ;)  Yes, it would be nice if Windows fixed MSI but after 10 years we know damn well they won't and 99% of operating systems used by customers today are impacted by this.  

Perhaps we could get a warning if the MSI contains UI tables but does not contain any ProgressText data with a message telling the developer where to read about this hot mess and what they might choose to do to make it better.
